src/tests/testCacheManager.cc

```diff
@@ -66,3 +75,150 @@ testCacheManager::testRegister()
     CPPUNIT_ASSERT_EQUAL(1,(int)sentry->flags);
 }
 
+void
+testCacheManager::testParseUrl()
+{
+    auto *mgr = static_cast<CacheManagerInternals *>(CacheManager::GetInstance());
+    CPPUNIT_ASSERT(mgr != nullptr);
+
+    std::vector<AnyP::ProtocolType> validSchemes = {
+        AnyP::PROTO_CACHE_OBJECT,
+        AnyP::PROTO_HTTP,
+        AnyP::PROTO_HTTPS,
+        AnyP::PROTO_FTP
+    };
+
+    AnyP::Uri mgrUrl;
+    mgrUrl.host("localhost");
+    mgrUrl.port(3128);
+
+    const std::vector<const char *> magicPrefixes = {
+        "/",
+        "/squid-internal-mgr/"
+    };
+
+    const std::vector<const char *> validActions = {
+        "",
+        "menu"
+    };
+
+    const std::vector<const char *> invalidActions = {
+        "INVALID" // any unregistered name
+    };
+
+    const std::vector<const char *> validParams = {
+        "",
+        "?",
+        "?&",
+        "?&&&&&&&&&&&&",
+        "?foo=bar",
+        "?0123456789=bar",
+        "?foo=bar&",
+        "?foo=bar&&&&",
+        "?&foo=bar",
+        "?&&&&foo=bar",
+        "?&foo=bar&",
+        "?&&&&foo=bar&&&&",
+        "?foo=?_weird?~`:[]stuff&bar=okay&&&&&&",
+        "?intlist=1",
+        "?intlist=1,2,3,4,5",
+        "?string=1a",
+        "?string=1,2,3,4,z",
+        "?string=1,2,3,4,[0]",
+        "?intlist=1,2,3,4,5&string=1,2,3,4,y"
+    };
+
+    const std::vector<const char *> invalidParams = {
+        "?foo",
+        "?/foo",
+        "?foo/",
+        "?foo=",
+        "?foo=&",
+        "?=foo",
+        "? foo=bar",
+        "? &",
+        "?& ",
+        "?=&",
+        "?&=",
+        "? &&&",
+        "?& &&",
+        "?&& &",
+        "?=&&&",
+        "?&=&&",
+        "?&&=&"
+    };
+
+    const std::vector<const char *> validFragments = {
+        "",
+        "#",
+        "##",
+        "#?a=b",
+        "#fragment"
+    };
+
+    unsigned success = 0;
+    for (const auto &scheme : validSchemes) {
+        mgrUrl.setScheme(scheme);
+
+        for (const auto *magic : magicPrefixes) {
+
+            // all schemes except cache_object require magic path prefix bytes
+            if (scheme != AnyP::PROTO_CACHE_OBJECT && strlen(magic) <= 2)
+                continue;
+
+            /* Check the parser accepts all the valid cases */
+
+            for (const auto *action : validActions) {
+                for (const auto *param : validParams) {
+                    for (const auto *frag : validFragments) {
+                        try {
+                            SBuf bits;
+                            bits.append(magic);
+                            bits.append(action);
+                            bits.append(param);
+                            bits.append(frag);
+                            mgrUrl.path(bits);
+
+                            (void)mgr->ParseUrl(mgrUrl);
+                            CPPUNIT_ASSERT(++success);
```

Comment:
It is not clear why this assert inside a loop is needed. And there is another assert like that below. Remove both?

Reply:
These are for cppuinit - to mark a check being done.

Reply:
> These are for cppuinit - to mark a check being done.

I do not understand what you mean by "to mark a check being done". The CppUnit [documentation](http://cppunit.sourceforge.net/doc/cvs/group___assertions.html) does not seem to imply that CPPUNIT_ASSERT() has some "marking" side effects (or any side effects), but that documentation is rather ... limited, so perhaps there is more to this assertion than meets the eye.

What would change in test logic and/or output if this CPPUNIT_ASSERT() line (with an never-failing condition) is removed?

Reply:
cppunit outputs a '.' character to visually mark successful checks in the logs and keep-alive any shells, network I/O and similar. These are useful to identify that the tested code is actually completing. It can take a long time on slow machines or environments running the processing at very low priority levels.

Reply:
> cppunit outputs a '.' character to visually mark successful checks

AFAICT, CPPUNIT_ASSERT() has nothing to do with progress reporting: CppUnit logs `.` at the beginning of each _test_, not after successful assertions. I cannot find any evidence that removing these always-true CPPUNIT_ASSERT()s will harm progress reporting (or anything else). Can you?

```C++
define CPPUNIT_ASSERT(condition)                                                 \
  ( CPPUNIT_NS::Asserter::failIf( !(condition),                                   \
                                 CPPUNIT_NS::Message( "assertion failed",         \
                                                      "Expression: " #condition), \
                                 CPPUNIT_SOURCELINE() ) )
```

```C++
void 
Asserter::failIf( bool shouldFail, 
                  const Message &message, 
                  const SourceLine &sourceLine )
{
  if ( shouldFail )
    fail( message, sourceLine );
}
```

```
$ git grep -F2 '"."' src
src/cppunit/TextTestProgressListener.cpp-TextTestProgressListener::startTest( Test * )
src/cppunit/TextTestProgressListener.cpp-{
src/cppunit/TextTestProgressListener.cpp:  stdCOut() << ".";
src/cppunit/TextTestProgressListener.cpp-  stdCOut().flush();
src/cppunit/TextTestProgressListener.cpp-}
--
src/cppunit/TextTestResult.cpp-{
src/cppunit/TextTestResult.cpp-  TestResultCollector::startTest (test);
src/cppunit/TextTestResult.cpp:  stdCOut() << ".";
src/cppunit/TextTestResult.cpp-}
src/cppunit/TextTestResult.cpp-
```

Reply:
Since you are pushing this 4a1b8de removes the ++success asserts for now. They can be added back if/when actually needed.